src/ros2cs/ros2cs_core/Node.cs

```diff
@@ -13,271 +13,373 @@
-    }
+        /// <summary>
+        /// Handle to the rcl node
+        /// </summary>
+        internal IntPtr Handle { get; private set; } = IntPtr.Zero;
 
-    /// <summary> Finalizer supporting IDisposable model </summary>
-    ~Node()
-    {
-      DestroyNode();
-    }
+        /// <summary>
+        /// Handle to the rcl node options
+        /// </summary>
+        private IntPtr Options = IntPtr.Zero;
 
-    /// <summary> Release managed and native resources. IDisposable implementation </summary>
-    public void Dispose()
-    {
-      DestroyNode();
-    }
+        /// <summary>
+        /// Context associated with this instance.
+        /// </summary>
+        private readonly Context ROSContext;
 
-    /// <summary> "Destructor" supporting IDisposable model </summary>
-    /// <description> Disposes all subscriptions and publishers and clients before finilizing node </description>
-    internal void DestroyNode()
-    {
-      lock (mutex)
-      {
-        if (!disposed)
+        /// <summary>
+        /// Lock used to allow thread safe access to node primitives.
+        /// </summary>
+        private readonly object Lock = new object();
+
+        /// <remarks>
+        /// This collection is thread safe.
+        /// </remarks>
+        /// <inheritdoc/>
+        public IReadOnlyCollection<IPublisherBase> Publishers { get; private set; }
+
+        private readonly HashSet<IRawPublisher> CurrentPublishers = new HashSet<IRawPublisher>();
+
+        /// <remarks>
+        /// This collection is thread safe.
+        /// </remarks>
+        /// <inheritdoc/>
+        public IReadOnlyCollection<ISubscriptionBase> Subscriptions { get; private set; }
+
+        private readonly HashSet<IRawSubscription> CurrentSubscriptions = new HashSet<IRawSubscription>();
+
+        /// <remarks>
+        /// This collection is thread safe.
+        /// </remarks>
+        /// <inheritdoc/>
+        public IReadOnlyCollection<IServiceBase> Services { get; private set; }
+
+        private readonly HashSet<IRawService> CurrentServices = new HashSet<IRawService>();
+
+        /// <remarks>
+        /// This collection is thread safe.
+        /// </remarks>
+        /// <inheritdoc/>
+        public IReadOnlyCollection<IClientBase> Clients { get; private set; }
+
+        private readonly HashSet<IRawClient> CurrentClients = new HashSet<IRawClient>();
+
+        /// <summary>
+        /// Create a new instance.
+        /// </summary>
+        /// <remarks>
+        /// The caller is responsible for adding the instance to <paramref name="context"/>.
+        /// This action is not thread safe.
+        /// </remarks>
+        /// <param name="name"> Name of the node. </param>
+        /// <param name="context"> Context to associate with. </param>
+        /// <exception cref="ObjectDisposedException"> If <paramref name="context"/> is disposed. </exception>
+        internal Node(string name, Context context)
```

Comment:
Exposing the `Node` constructor would allow users to create nodes without the context knowing which in turn would allow
disposing the context before disposing the nodes.
While Iam not sure if it is allowed I think it is not.